hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java

```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 		}
 	}
 
+	private <T extends AST> List<T> findAllNodes(AST node, Class<T> clazz) {
+		ArrayList<T> result = new ArrayList<>();
+		findAllNodesCore( node, clazz, result );
+		return result;
+	}
+
+	private <T extends AST> void findAllNodesCore(AST node, Class<T> clazz, List<T> found) {
+		if ( clazz.isAssignableFrom( node.getClass() ) ) {
+			found.add( (T) node );
+		}
+		if ( node.getFirstChild() != null ) {
+			findAllNodesCore( node.getFirstChild(), clazz, found );
+		}
+		if ( node.getNextSibling() != null ) {
+			findAllNodesCore( node.getNextSibling(), clazz, found );
+		}
+	}
+
+	private Set<String> findQueryReferencedTables(QueryNode query) {
+		if ( !walker.getSessionFactoryHelper()
+				.getFactory()
+				.getSessionFactoryOptions()
+				.isOmitJoinOfSuperclassTablesEnabled() ) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug( "Finding of query referenced tables is skipped because the feature is disabled by " +
+							"configuration. See " + AvailableSettings.OMIT_JOIN_OF_SUPERCLASS_TABLES );
+			}
+			return null;
+		}
+
+		if ( walker.getEnabledFilters() != null && !walker.getEnabledFilters().isEmpty() ) {
+			LOG.debug( "Finding of query referenced tables is skipped because filters are enabled." );
+			return null;
+		}
+
+		if ( LOG.isDebugEnabled() ) {
+			LOG.debug( TokenPrinters.REFERENCED_TABLES_PRINTER.showAsString(
+					query,
+					"Tables referenced from query nodes:"
+			) );
+		}
+
+		Set<String> result = new HashSet<>();
+
+		// Find tables referenced by FromReferenceNodes
+		List<FromReferenceNode> fromReferenceNodes = findAllNodes( query, FromReferenceNode.class );
+		for ( FromReferenceNode node : fromReferenceNodes ) {
+			String[] tables = node.getReferencedTables();
+			if ( tables != null ) {
+				Arrays.stream( tables )
+						.filter( Objects::nonNull )
```

Comment:
Can't we filter the nulls before adding them?

Reply:
Yes, `getReferencedTables()` should not returns any null elements in the array, null check is useless. Removed.